Init or upgrade schema from openQA

aaannz opened the pull request:
Preparing the database upgrade is the same as before, but manual
initdb resp. upgradedb call to perform initialization resp. upgrade
is not needed.
Upon first connect, openQA will check if database contain schema version.
If none is found, db is initialized. Else it is compared to current available
schema version and if not matching, upgrade is done automatically.

okurz commented on lib/OpenQA/Schema.pm:
```diff
@@ -14,40 +14,107 @@
+                    my ($storage, $dbh, @args) = @_;
+                    $dbh->do("PRAGMA foreign_keys = OFF;");
+                });
+        }
```
could you extract this part for sqlite into a helper sub? it is used twice -> DRY :-)

aaannz replied:
Twice? One ime is `PRAGMA synchonous = off` and second time `PRAGMA foreign_keys`. Or you mean to hide the same bolierplate?

okurz replied:
right, I *did* oversee the important difference :-) So I would extract a helper sub with argument

okurz commented:
see inline comment. could this have any effect on local testing with morbo or the runtime of local tests?

aaannz commented:
I have slow start today, what do you mean by 'inline comment'? Anyway, there should be no effect on local testing and etc.

okurz commented:
with "inline comment" I mean the comment about extracting a sub. I just write this as a "cover message" in the main discussion thread to show that I reviewed all infile coments. I want to prevent people to actually work on their changes, pushing an update while I am still reviewing the in file content and then either the comment is lost (as belonging to an "outdated diff") or the reviewee has to do another change when he could have done it all in one.

AdamWill commented:
What happens if the deployment check fails?

Can we test the upgrade process?

aaannz commented:
When $dh->install or $dh->upgrade fails I expect it to die.

I looked briefly into testing upgrade, but so far I didn't find good solution. For upgrade I need to prepare it and we are reusing same dbicdh directory.
Hmm.. as I'm writing this now, I kind of remember DH supports initializing db to specific schema version. In theory I can init it to version - 1 and test upgrade.
However without some meaningful data it would be more of a sanity check.

coolo commented:
I have seriously bad feelings about adding this into Schema.pm - it just shouts "dead lock" at me. How about adding the upgrade hook into script/openqa ?

aaannz commented:
Hmm.. that's an interesting idea. Indeed I had a few issues, not dead locks per se, but due to numerous code duplications (OpenQA::Test::Database, t/deploy.t) it took quite effort to stabilize.

aaannz commented:
I removed the automatic deployment check on db connect, instead it must be called manually if needed. I put this call to `startup` sub, since in openqa script I didn't want to add arguments parsing which I would had to do at least for `openqa test` call.
Also I added `disconnect_db` subroutine so I can clean db and reinitialize it in deployment test.

okurz commented:
LGTM. Feel free to bump the test threshold to 60.9, the result of your test, and auto-LGTM
